### PR TITLE
[Don't merge this yet] Merge geospatial topic branch into master

### PR DESCRIPTION
**Pull request by wolffcm**

### CODE REVIEW

File: src/ee/common/NValue.hpp [hunk 1]
```diff
@@ -1473,6 +1507,9 @@ class NValue {
             streamTimestamp(value);
             break;
         }
+        case VALUE_TYPE_POINT: {
+            value << getPoint().toString(); break;
+        }
```
Comment (paulmartel) [inline src/ee/common/NValue.hpp]: MISSING: case VALUE_TYPE_GEOGRAPHY:

File: src/ee/common/NValue.hpp [hunk 2]
```diff
@@ -1720,11 +1766,11 @@ class NValue {
...
-        int64_t lhsValue, rhsValue;
-        lhsValue = static_cast<int64_t>(getSmallInt());
-        rhsValue = rhs.castAsBigIntAndGetValue();
-        return compareValue<int64_t>(lhsValue, rhsValue);
-    }
+            int64_t lhsValue, rhsValue;
+            lhsValue = static_cast<int64_t>(getSmallInt());
+            rhsValue = rhs.castAsBigIntAndGetValue();
+            return compareValue<int64_t>(lhsValue, rhsValue);
+        }
```
Comment (paulmartel) [inline src/ee/common/NValue.hpp]: revert strange extra indent

File: src/ee/common/NValue.hpp [hunk 3]
```diff
@@ -1955,8 +2001,42 @@ class NValue {
                  valueToString(rhs.getValueType()).c_str(),
                  valueToString(getValueType()).c_str());
         throw SQLException(SQLException::
-                           data_exception_most_specific_type_mismatch,
-                           message);
+                data_exception_most_specific_type_mismatch,
+                message);
+    }
+
+    int comparePointValue (const NValue rhs) const {
```
Comment (paulmartel) [inline src/ee/common/NValue.hpp]: Missing migration to use reference (at least for consistency): const NValue& rhs

File: src/ee/common/NValue.hpp [hunk 4]
```diff
@@ -1955,8 +2001,42 @@ class NValue {
...
+                << " cannot be cast for comparison to type "
+                << valueToString(getValueType());
+            throw SQLException(SQLException::data_exception_most_specific_type_mismatch,
+                               oss.str().c_str());
+            // Not reached
+            return 0;
+        }
+    }
+
+    int compareGeographyValue (const NValue rhs) const {
```
Comment (paulmartel) [inline src/ee/common/NValue.hpp]: Missing migration to use reference (at least for consistency): const NValue& rhs

File: src/ee/common/NValue.hpp [hunk 5]
```diff
@@ -2256,6 +2336,12 @@ class NValue {
         return retval;
     }
 
+    static NValue getUninitializedAllocatedValue(ValueType type, size_t size, Pool* stringPool) {
```
Comment (paulmartel) [inline src/ee/common/NValue.hpp]: Suggestion:
Move these two lines into the only caller -- getUninitializedTempGeographyValue --
and eliminate this function (again).

File: src/ee/common/NValue.hpp [hunk 6]
```diff
@@ -2851,8 +2983,22 @@ inline void NValue::deserializeFromAllocateForStorage(ValueType type, SerializeI
...
-        const char *str = (const char*) input.getRawPointer(length);
-        createObjectPointer(length, str, tempPool);
+
+        if (type != VALUE_TYPE_GEOGRAPHY) {
+            const char *str = (const char*) input.getRawPointer(length);
+            createObjectPointer(length, str, tempPool);
+        }
+        else {
+            // Allocate a StringRef, but pass a null pointer so we don't
+            // copy data into it yet.
```
Comment (paulmartel) [inline src/ee/common/NValue.hpp]: Simplify/optimize/make more consistent with deserializeFrom by replacing below with:

                StringRef* sref = createObjectPointer(length, NULL, tempPool);
                GeographyValue::deserializeFrom(input, sref->getObjectValue(), length);

File: src/ee/common/NValue.hpp [hunk 7]
```diff
@@ -3192,10 +3357,17 @@ inline void NValue::hashCombine(std::size_t &seed) const {
     }
     case VALUE_TYPE_DECIMAL:
         getDecimal().hash(seed);
+        return;        getDecimal().hash(seed); break;
```
Comment (paulmartel) [inline src/ee/common/NValue.hpp]: dead garbage after return.

File: src/ee/common/NValue.hpp (hunk 1 above)
Comment (wolffcm) [inline src/ee/common/NValue.hpp]: Good catch.  John Crawford filed a ticket for this.

**Pull request merged**